Sources/NIOSSL/SSLContext.swift

```diff
@@ -211,6 +203,78 @@ private func clientPSKCallback(ssl: OpaquePointer?,
+    let contextManager = parentSwiftContext.customContextManager!
+
+    switch contextManager.state {
+    case .none:
+        // Construct extension values
+        let cServerHostname = CNIOBoringSSL_SSL_get_servername(ssl, TLSEXT_NAMETYPE_host_name)
+        let serverHostname = cServerHostname.map { String(cString: $0) }
+        let values = NIOSSLClientExtensionValues(serverHostname: serverHostname)
+
+        // Load the attached connection so we can resume handshake when future resolves
+        let connection = SSLConnection.loadConnectionFromSSL(ssl)
+
+        // Issue call to get a new ssl context
+        // This is a safe force unwrap because the context callback
+        // is always saved in the case where we register set_cert_cb
+        contextManager.loadContext(values: values, on: connection)
+
+        // We must return a negative value to suspend the handshake
+        return -1
+    case .complete(let sslContextCallbackResult):
+        let userChosenContext: NIOSSLContext
+        switch sslContextCallbackResult {
+        case .success(let value):
+            userChosenContext = value
+        case .failure:
+            return SSL_TLSEXT_ERR_NOACK
+        }
+
+        // The user may or may not retain a reference to the NIOSSLContext
+        // that they return here, if they provide one. The NIOSSLContext finalizer
+        // explicitly decrements the reference count on the underlying SSL_CTX.
+        //
+        // However, SSL_set_SSL_CTX explicitly increases the reference count and
+        // will decrement the count when the SSL* object is destroyed, thus finally
+        // freeing the raw SSL_CTX object, even if the swift reference is orphaned.
+        //
+        // As such, no memory should be leaked or freed too early in the event that
+        // the user fails to retain their created and return NIOSSLContext. In the
+        // event that they do retain it, the reference count will remain at at-least
+        // 1 and thus they can re-use the same context for multiple connections.
+        let nativeContextValue = withExtendedLifetime(userChosenContext) {
+            return CNIOBoringSSL_SSL_set_SSL_CTX(ssl, userChosenContext.sslContext)
+        }
+
+        // CNIOBoringSSL_SSL_set_SSL_CTX should return a pointer to the very same context it was
+        // told to assign on success. On error, this will not happen, so we check for equality
+        // here and handle any such errors by signaling the connection to fail.
+        if nativeContextValue != userChosenContext.sslContext {
+            return SSL_TLSEXT_ERR_NOACK
+        }
+        
+        // We must return 1 to signal a successful load of the new context
+        return 1
+    case .loading:
+        preconditionFailure("""
+            SSL_CTX_set_cert_cb was executed in a pending state.
+            This should not be possible, please file an issue.
+        """)
```

Comment:
I think this is gonna be too aggressive. We may, for other reasons, end up calling `doHandshake` again (e.g. because more data arrives). We should tolerate being called again, and just returning negative values.